File: 583-delete-operation-for-two-strings/583-delete-operation-for-two-strings.py
class Solution:
    def minDistance(self, w1: str, w2: str) -> int:
        m,n = len(w1),len(w2)
        @cache
        def lcs(i,j):
            if i<0 or j<0:return 0
            ans = lcs(i-1,j)
            ans = max(ans,lcs(i,j-1))
            if w1[i]==w2[j]:
                ans = max(ans,1+lcs(i-1,j-1))
            return ans
                
        
        temp = lcs(m-1,n-1)
        
        return (m+n)-2*temp

File: 583-delete-operation-for-two-strings/583-delete-operation-for-two-strings.cpp
class Solution {
public:
    int DP(int a, int b, string& word1, string& word2, vector<vector<int>>& dp_arr){
        if(dp_arr[a][b]!=-1)return dp_arr[a][b];
        if(a==0 || b==0)return dp_arr[a][b]=a+b;
        else if(word1[a-1]==word2[b-1]){
            return dp_arr[a][b]=
                DP(a-1, b-1, word1, word2, dp_arr);
        }
        return dp_arr[a][b]=1+min(
            DP(a-1, b, word1, word2, dp_arr),
            DP(a, b-1, word1, word2, dp_arr)
        );
    }
    int minDistance(string word1, string word2) {
        vector<vector<int> > dp_arr(word1.size()+1 , vector<int> (word2.size()+1, -1));
        return DP(word1.length(), word2.length(), word1, word2, dp_arr);
        
    }
};